dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageChecker.java

```diff
@@ -54,87 +60,119 @@
+    return unresolvedMethodReferences;
   }
 
   /**
    * Checks the availability of the methods through the jar files, and lists the unavailable
    * methods.
    *
-   * @param paths jar files to search for the availability of the methods
-   * @param methodReferences methods to search for with the jar files
+   * @param jarFilePaths absolute paths to the jar files to search for the methods
+   * @param methodReferences methods to search for within the jar files
    * @return list of methods that are not found in the jar files
    */
-  static List<FullyQualifiedMethodSignature> resolvedMethodReferences(List<Path> paths,
+  static List<FullyQualifiedMethodSignature> findUnresolvedReferences(List<Path> jarFilePaths,
       List<FullyQualifiedMethodSignature> methodReferences) {
     List<FullyQualifiedMethodSignature> unresolvedMethods = new ArrayList<>();
 
-    // Creates chain of ClassPath items in the same order as paths
+    // Creates chain of ClassPath items in the same order as jarFilePaths
     ClassPath classPath = null;
-    for (Path path : paths) {
-      String pathFileName = path.toFile().getAbsolutePath();
-      classPath = new ClassPath(classPath, pathFileName);
+    for (Path absolutePathToJar : jarFilePaths) {
+      classPath = new ClassPath(classPath, absolutePathToJar.toString());
     }
+    Set<String> classesNotFound = new HashSet<>();
+    Set<FullyQualifiedMethodSignature> availableMethodsInJars = new HashSet<>();
     SyntheticRepository repository = SyntheticRepository.getInstance(classPath);
 
     for (FullyQualifiedMethodSignature methodReference : methodReferences) {
+      String className = methodReference.getClassName();
+
+      // Case 1: we know that the class doesn't exist in the jar files
+      if (classesNotFound.contains(className)) {
+        unresolvedMethods.add(methodReference);
+        continue;
+      }
+      // Case 2: we know that the class and method exist in the jar files
+      if (availableMethodsInJars.contains(methodReference)) {
+        continue;
+      }
+
+      // Case 3: we need to check the availability through repository
       try {
-        JavaClass javaClass = repository.loadClass(methodReference.getClassName());
-        MethodSignature methodSignature = methodReference.getMethodSignature();
-        boolean methodFound = false;
-        // Opportunity to get better performance to create a cache (from class name to method list)
-        for (Method method : javaClass.getMethods()) {
-          String signature = method.getSignature();
-          if (method.getName().equals(methodSignature.getMethodName())
-              && signature.equals(methodSignature.getDescriptor())) {
-            methodFound = true;
-            break;
-          }
-        }
-        if (! methodFound) {
+        JavaClass javaClass = repository.loadClass(className);
+        List<FullyQualifiedMethodSignature> availableMethodsOnClass = listMethodsOnClass(javaClass);
```

Comment:
This seems like something that is quite expensive to do for every lookup, because this is an O(n) operation where n = number of methods in the class. This seems like something that should be cached.

Reply:
Knuth!

Reply:
@garrettjonesgoogle cache `availableMethodsInJars` has been added already https://github.com/GoogleCloudPlatform/cloud-opensource-java/pull/121/files/ae3f18af82ffb34a423dbe9cdbdeaf7641eaca51..5ecf5ca270c8d4dbac5e6e610d7ea176f36d5d54#diff-cb2eefbf6e8ca674c445c1c7e1b519edR140

Reply:
Ahh ok.